### PR TITLE
[lld][ELF] Add basic TLSDESC support for LoongArch

### PR DESCRIPTION
**Pull request by wangleiat**
LoongArch does not yet implement transition from TLSDESC to LE/IE,
so TLSDESC dynamic relocation needs to be generated for each desc,
which is ultimately handled by the dynamic linker.

The test cases reference RISC-V: #79239

### CODE REVIEW

**Comment by wangleiat** [comment]
@xen0n

**Review by xen0n — COMMENTED** [review]
I haven't manually verified the test case yet but it seems okay.

File: lld/ELF/Relocations.cpp [hunk 1]
```diff
@@ -1303,6 +1303,19 @@ static unsigned handleTlsRelocation(RelType type, Symbol &sym,
 
   if (config->emachine == EM_MIPS)
     return handleMipsTlsRelocation(type, sym, c, offset, addend, expr);
+
+  // LoongArch does not yet implemented TLSDESC transition of DESC to LE/IE.
+  // Generates TLSDESC dynamic relocation for the symbol that is handled by the
+  // dynamic linker.
```
[the remark this answers is not in the document]
Comment (wangleiat) [reply lld/ELF/Relocations.cpp]: Thanks.

File: lld/test/ELF/loongarch-tlsdesc.s [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+.globl a
+.zero 8
+a:
+.zero 2039  ## Place b at 0x7ff
+b:
+.zero 1
+
+#--- c.s
+.section .tbss,"awT",@nobits
+#.tbss
```
Comment (MaskRay) [inline lld/test/ELF/loongarch-tlsdesc.s:132]: delete `.tbss`

File: lld/test/ELF/loongarch-tlsdesc.s [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+la.tls.desc $a0, a
+add $a1, $a0, $tp
+
+la.tls.desc $a0, b
+add $a2, $a0, $tp
+
+la.tls.desc $a0, c
+add $a3, $a0, $tp
+
+#.section .tbss
```
Comment (MaskRay) [inline lld/test/ELF/loongarch-tlsdesc.s:121]: delete the comment

File: lld/test/ELF/loongarch-tlsdesc-gd-mixed.s [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# REQUIRES: loongarch
+# RUN: llvm-mc -filetype=obj -triple=loongarch64 %s -o %t.o
+# RUN: ld.lld -shared %t.o -o %t.so
+# RUN: llvm-readobj -r %t.so
```
Comment (MaskRay) [inline lld/test/ELF/loongarch-tlsdesc-gd-mixed.s:4]: Missing FileCheck

**Review by MaskRay — COMMENTED** [review]
The code change looks good. Do you have plan for TLSDESC to LE/IE optimization?

**Comment by MaskRay** [comment]
Consider mentioning #79239 in the description.

**Comment by wangleiat** [comment]
> The code change looks good. Do you have plan for TLSDESC to LE/IE optimization?

Sorry for the late reply due to the holiday.  No plans in the near future.
This implementation is just to allow lld to simply process `tlsdesc`.

**Comment by wangleiat** [comment]
> Consider mentioning #79239 in the description.

Thanks very much!

**Comment by wangleiat** [comment]
gentle ping.

**Review by SixWeining — APPROVED** [review]
LGTM.

File: lld/ELF/Relocations.cpp [hunk 5]
```diff
@@ -1297,6 +1297,18 @@ static unsigned handleTlsRelocation(RelType type, Symbol &sym,
 
   if (config->emachine == EM_MIPS)
     return handleMipsTlsRelocation(type, sym, c, offset, addend, expr);
+
+  // LoongArch does not yet implement transition from TLS DESC to LE/IE, so
```
Comment (MaskRay) [inline lld/ELF/Relocations.cpp:1301]: `TLS DESC` => `TLSDESC`

**Review by MaskRay — APPROVED** [review]

**Pull request merged**